### PR TITLE
Absorb c10d into the main cmake build

### PR DESCRIPTION
**Pull request by anderspapitto**

### CODE REVIEW

File: torch/lib/c10d/example/CMakeLists.txt [hunk 1]
```diff
@@ -1,3 +1,5 @@
 add_executable(allreduce allreduce.cpp)
 target_include_directories(allreduce PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/..)
 target_link_libraries(allreduce pthread c10d)
+
+add_dependencies(allreduce c10d)
```
Comment (pietern) [inline torch/lib/c10d/example/CMakeLists.txt]: Is this not taken care of by target_link_dependencies above?

File: torch/lib/THD/CMakeLists.txt [hunk 2]
```diff
@@ -67,9 +63,8 @@ IF(MPI_FOUND)
   MESSAGE(STATUS "MPI_LIBRARIES: ${MPI_LIBRARIES}")
 ENDIF()
 
-IF(GLOO_FOUND)
+IF(USE_GLOO AND USE_CUDA)
```
Comment (soumith) [inline torch/lib/THD/CMakeLists.txt]: I wont repeat the THD comments from #12775 again over here, have a look at that review

File: torch/lib/c10d/CMakeLists.txt [hunk 3]
```diff
@@ -127,19 +106,29 @@ target_compile_options(c10d PUBLIC
...
+
+if (USE_GLOO)
+  add_dependencies(c10d gloo)
+  target_link_libraries(c10d PUBLIC gloo)
+  if (USE_CUDA)
+    target_link_libraries(c10d PUBLIC gloo_cuda)
+  endif()
 endif()
 
+target_include_directories(c10d PUBLIC
```
Comment (soumith) [inline torch/lib/c10d/CMakeLists.txt]: if you want, add a TODO comment that this has to be cleaned up later.

**Review by soumith — APPROVED** [review]
approved for the c10d commit

**Pull request closed without merging**